pkg/proc/proc.go

```diff
@@ -523,21 +534,34 @@ func StepInstruction(dbp *Target) (err error) {
 	if ok, err := dbp.Valid(); !ok {
 		return err
 	}
-	thread.Breakpoint().Clear()
-	err = thread.StepInstruction()
-	if err != nil {
-		return err
-	}
-	err = thread.SetCurrentBreakpoint(true)
-	if err != nil {
+
+	if err := dbp.threadStepInstruction(thread, true); err != nil {
 		return err
 	}
+
 	if tg, _ := GetG(thread); tg != nil {
 		dbp.SetSelectedGoroutine(tg)
 	}
 	return nil
 }
 
+func (dbp *Target) threadStepInstruction(thread Thread, setbp bool) error {
+	bp := dbp.ThreadToBreakpoint(thread)
+	if bp.Breakpoint != nil {
+		if err := dbp.Process.ClearBreakpointFn(bp.Addr, bp.OriginalData); err != nil {
+			return err
+		}
+		defer dbp.Process.WriteBreakpointFn(bp.Addr)
+	}
+	if err := thread.StepInstruction(); err != nil {
```

Comment:
The way this works out is that it becomes far more inefficient in the gdbserial backend, the original code called the internal gdbserial.(*Thread).stepInstruction, which did not reload the thread registers (because we don't need to know in this situation), while this calls StepInstruction which will reload the threads.

Reply:
Perhaps that's a sign the method is doing too much. Why should `StepInstruction` have a side effect of reloading registers? I think a more proper place for that would be in the first call to `Thread.Registers` and then the result is memoized. Essentially calling and caching on demand instead of as a result of calling `StepInstruction`.

Reply:
> Perhaps that's a sign the method is doing too much

You are not wrong,  but the gdbserial backend was designed to always keep the regs field of Thread always updated. I did that because it was simpler but doing it when it's needed and memoizing is definitely the way forward.

However it's a rather complicated change so it should be done in a separate PR, rather than as part of an unrelated refactor. And either that goes first or this PR goes in without moving the thread resume logic out of the backends.

Reply:
Agreed.